rust_src/src/javascript.rs

```diff
@@ -0,0 +1,280 @@
+                let scope = &mut v8::HandleScope::with_context(runtime.v8_isolate(), context);
+                let context = scope.get_current_context();
+                let global = context.global(scope);
+                {
+                    let name = v8::String::new(scope, "lisp_invoke").unwrap();
+                    let func = v8::Function::new(scope, lisp_callback).unwrap();
+                    global.set(scope, name.into(), func.into());
+                }
+                {
+                    let name = v8::String::new(scope, "is_proxy").unwrap();
+                    let func = v8::Function::new(scope, is_proxy).unwrap();
+                    global.set(scope, name.into(), func.into());
+                }
+            }
+            {
+                runtime
+                    .execute(
+                        "prelim.js",
+                        "var lisp = new Proxy({}, {
+                get: function(o, k) {
+                   return function() {
+                       const modargs = [k.replaceAll('-', '_')];
+                          for (let i = 0; i < arguments.length; ++i) {
+                             if (is_proxy(arguments[i])) {
+                                 modargs.push(arguments[i]);
+                             } else {
+                                 modargs.push(JSON.stringify(arguments[i]));
+                             }
+                          }
+                       let result = lisp_invoke.apply(this, modargs);
+                       let retval = is_proxy(result) ? result : JSON.parse(result);
+                       return retval;
+                   }
+
+                }});",
+                    )
+                    .unwrap();
+            }
+
+            worker
+        });
+
+        EmacsJsRuntime::set_main(r, worker);
+        //(run-with-timer t 1 'js-tick-event-loop)
+
+        let cstr = CString::new("run-with-timer").expect("Failed to create timer");
+        let callback = CString::new("js-tick-event-loop").expect("Failed to create timer");
+        unsafe {
+            let fun = crate::remacs_sys::intern_c_string(cstr.as_ptr());
+            let fun_callback = crate::remacs_sys::intern_c_string(callback.as_ptr());
+            let mut args = vec![
+                fun,
+                crate::remacs_sys::Qt,
+                crate::remacs_sys::make_float(0.1),
+                fun_callback,
+            ];
+            crate::remacs_sys::Ffuncall(args.len().try_into().unwrap(), args.as_mut_ptr());
+        }
+    });
+
+    execute!(async move |mut w: deno_runtime::worker::MainWorker| {
+        if let Some(js) = additional_js {
+            w.execute(&js).unwrap();
+        } else {
+            let main_module = deno_core::ModuleSpecifier::resolve_url_or_path(filepath).unwrap();
+            w.execute_module(&main_module).await.unwrap();
+        }
+
+        w
+    });
+
+    crate::remacs_sys::Qnil
+}
+
+#[lisp_fn]
+pub fn js_tick_event_loop() -> LispObject {
```

Comment:
can you elaborate how this will work? Can you also list what are the performance impacting decisions made in this PR and whether you think they can be avoided?

Reply:
When we initialize the runtime, we set up a timer to call js_tick_event_loop. When that timer fires, we call tick_js!(), which fires execute!(). execute! blocks the current thread (the elisp thread) while javascript is running within the context of the tokio runtime. Once that returns, we will exit our timer callback. This is expected to be a fast operation when no pending javascript is awaiting execution.